core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/AbstractProgressStep.java

```diff
@@ -0,0 +1,127 @@
+package uk.ac.bbsrc.tgac.miso.core.data.workflow;
+
+import java.io.Serializable;
+
+import javax.persistence.Column;
+import javax.persistence.Embeddable;
+import javax.persistence.EmbeddedId;
+import javax.persistence.Entity;
+import javax.persistence.Inheritance;
+import javax.persistence.InheritanceType;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import javax.persistence.Table;
+
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.impl.ProgressImpl;
+
+@Entity
+@Table(name = "WorkflowProgressStep")
+@Inheritance(strategy = InheritanceType.JOINED)
+public abstract class AbstractProgressStep implements ProgressStep {
+  @EmbeddedId
+  private ProgressStepId id;
+
+  public ProgressStepId getId() {
+    return id;
+  }
+
+  public void setId(ProgressStepId id) {
+    this.id = id;
+  }
+
+  @Override
+  public Progress getProgress() {
+    return id.getProgress();
+  }
+
+  @Override
+  public void setProgress(Progress progress) {
+    this.id.setProgress(progress);
+  }
+
+  @Override
+  public int getStepNumber() {
+    return id.getStepNumber();
+  }
+
+  @Override
+  public void setStepNumber(int stepNumber) {
+    this.id.setStepNumber(stepNumber);
+  }
+
+  @Override
+  public int hashCode() {
+    final int prime = 31;
+    int result = 1;
+    result = prime * result + ((id == null) ? 0 : id.hashCode());
+    return result;
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (this == obj)
+      return true;
+    if (obj == null)
+      return false;
+    if (getClass() != obj.getClass())
+      return false;
+    AbstractProgressStep other = (AbstractProgressStep) obj;
+    if (id == null) {
+      if (other.id != null)
+        return false;
+    } else if (!id.equals(other.id))
+      return false;
+    return true;
+  }
+
+  @Override
+  public int compareTo(ProgressStep o) {
+    return Integer.compare(id.getStepNumber(), o.getStepNumber());
+  }
+
+  @Embeddable
+  public static class ProgressStepId implements Serializable {
+    private static final long serialVersionUID = 1L;
+
+    @ManyToOne(targetEntity = ProgressImpl.class)
+    @JoinColumn(name = "workflowProgressId")
+    private transient Progress progress;
+
+    @Column(nullable = false)
+    private int stepNumber;
+
+    Progress getProgress() {
+      return progress;
+    }
+
+    void setProgress(Progress progress) {
+      this.progress = progress;
+    }
+
+    int getStepNumber() {
+      return stepNumber;
+    }
+
+    void setStepNumber(int stepNumber) {
+      this.stepNumber = stepNumber;
+    }
+
+    @Override
+    public int hashCode() {
+      final int prime = 31;
+      int result = 1;
+      result = prime * result + stepNumber;
+      return result;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```

Comment:
`equals` and `hashcode` are only comparing `stepNumber`, which is half of the key. Should compare the Progress objects or their IDs too

Reply:
:+1: 

`AbstractProgressStep#equals` now compares `Progress` objects.

I've modified `ProgressImpl#equals` to *not* compare `ProgressStep` objects, so that we don't get an infinite loop.